Add blog post about cache support collab (#615)
trinodb/trino.io: _posts/2024-02-26-cache-refresh.md

---
layout: post
title:  "A cache refresh for Trino"
author: Manfred Moser
image: /assets/blog/trino-cache-refresh.png
excerpt_separator: <!--more-->
show_pagenav: true
---

> Thinking about our recent work on caching in Trino reminds me of he famous

Reviewer: ```suggestion
Thinking about our recent work on caching in Trino reminds me of the famous
```

> saying [There are only two hard things in computer science: cache invalidation
> and naming things](https://www.karlton.org/2017/12/naming-things-hard/). Well,

Reviewer: ```suggestion
saying, ["There are only two hard things in computer science: cache invalidation
and naming things](https://www.karlton.org/2017/12/naming-things-hard/)." Well,
```

in the Trino community we know all about caching and naming. With the recent
[Trino 439 release]({{site.url}}/docs/current/release/release-439.html), caching
from object storage file systems got a refresh. Catalogs using the Delta Lake,
Hive, and Iceberg connectors now get to access performance benefits from the new
Alluxio-powered file system caching.

<!--more-->

## In the past

> So how did we get here? A long, long time ago, Quoble open-sourced a [light

Reviewer 2: typo: Qubole

Author: Thanks .. fixed

docs. Mateusz also added support for OpenTelemetry.

[Manish Malhorta](https://github.com/osscm) and [Amogh
Margoor](https://github.com/amoghmargoor) sent a PR for Iceberg support. They
were also about to add Hive support, when [Raunaq
Morarka](https://github.com/raunaqmorarka) beat them and submitted that PR.

After some final clean up, [Cole Bowden](https://github.com/colebow) and [Martin
Traverso](https://github.com/martint) got the release notes together and shipped
[Trino 439]({{site.url}}/docs/current/release/release-438.html)! Now you can use
it too.

## Using file system caching

There are only a few relatively simple steps to add file system caching to your
catalogs that use Delta Lake, Hive, or Iceberg connectors:

* Provision fast local file system storage on all your Trino cluster nodes. How
  you do that depends on your cluster provisioning.
* Enable file system caching and configure the cache location, for example at
  `/tmp/trino-cache` on the nodes for in your catalog properties files.

```
fs.cache.enabled=true
fs.cache.directories=/tmp/trino-cache
```

After a cluster restart, file system caching is active for the configured

> catalogs. You can  and you can tweak it with [further, optional configuration

Reviewer 3: double space here?

Author: 🦅 👁️ @dain

---
layout: post
title:  "A cache refresh for Trino"
author: Manfred Moser
image: /assets/blog/trino-cache-refresh.png
excerpt_separator: <!--more-->
show_pagenav: true
---

Leon Bambrick once said "There are only two hard things in computer science:
cache invalidation, naming things, and off-by-one errors" improving the initial

> wisdom from [Phil Karlton](https://www.karlton.org/2017/12/naming-things-hard/).

Reviewer: I'm a little uncertain about this quote and citing two different individuals, as well as linking to this kinda sketchy website. I'm not familiar with the person we're referencing, and I don't think there's an expectation for them to be well-known names, so proper writing would need to introduce or explain background on them.

But something like, "There's a famous proverb: 'There are only two hard things...'" without attribution is actually fair game, as the quote really is a proverb at this point and [thus doesn't need to be attributed](https://style.mla.org/citing-famous-sayings/).

The Leon Bambrick variation isn't as much a proverb, and I'd think we can just drop the "off by one errors" portion of the quote to avoid the messiness of attributing it to a not-super-notable individual and needing to explain who they are.

Author: The website is not sketchy at all.. just old. And the two authors are both famous and I think we want to at least attribute to the Karlton site.. and it has lots more links to related stuff. I chose the Bambrick quote because its funny..  so it makes for a light entry to the blog post

Well, in the Trino community we know all about caching, naming, off-by-one
errors, and improving on original wisdom. With the recent [Trino 439

> release]({{site.url}}/docs/current/release/release-438.html) caching from object

Reviewer: ```suggestion
release]({{site.url}}/docs/current/release/release-438.html), caching from object
```

> storage file systems got a refresh. Users of catalogs using the Delta Lake,

Reviewer: "Users of catalogs using" sounds kind of redundant, I'd reword.

Author: done

Hive, and Iceberg connectors now get to enjoy performance benefits from the new
Alluxio-powered file system caching. And all that as a replacement from the now

> removed object storage caching with RubiX.

Reviewer: This last sentence is a fragment - maybe say, "In case you're wondering about the old caching we had in Hive, this is replacing..."

Author: I am removing the sentence.

<!--more-->

## In the past

> So how did we get here? A long, long time ago Quoble open sourced a [light

Reviewer: A 3-for-1 grammar nitpick may be my new record
```suggestion
So how did we get here? A long, long time ago, Qubole open-sourced a [light
```

light-weight data caching framework called
RubiX](https://github.com/qubole/rubix). The library was integrated into the

> Trino Hive connector and enabled [Hive connector storage

Reviewer: ```suggestion
Trino Hive connector, and it enabled [Hive connector storage
```

> caching]({{site.url}}/docs/438/connector/hive-caching.html). However, over time

Reviewer: We need a comma after "over time," so let's change "However" to "But" to simplify the flow and not have too many commas breaking things up.
```suggestion
caching]({{site.url}}/docs/438/connector/hive-caching.html). But over time,
```

Author: Done

any open source project without active maintenance becomes stale. And like a

> stale cache, a stale open source project can cause issues, or at a minimum

Reviewer: ```suggestion
stale cache, a stale open source project can cause issues, or
```

becomes outdated and unsuitable for modern use. Staleness and other problems

> meant that RubiX had to go.

Reviewer: I think we could phrase this a little lighter, or even insert some humor. Whether that's, "though RubiX had once served Trino well, it was its time for it to go," or referencing it going to the great software repo in the sky, I leave to you.

Author: Updated

## Making progress

> Time moved on and Trino now includes powerful connectors for the modern

Reviewer: ```suggestion
Catching back up to 2024, Trino now includes powerful connectors for the modern
```
Or something similar, "time moved on" sounds a little melodramatic.

  <div class="row">
    <div class="col-sm">
      <a href="{{site.url}}/docs/current/connector/delta-lake.html" target="_blank">
        <img src="{{site.url}}/assets/images/logos/delta-lake.png" title="Delta Lake connector">
      </a>
    </div>
    <div class="col-sm">
      <a href="{{site.url}}/docs/current/connector/hudi.html" target="_blank">
        <img src="{{site.url}}/assets/images/logos/apache-hudi.png" title="Hudi connector">
      </a>
    </div>
    <div class="col-sm">
      <a href="{{site.url}}/docs/current/connector/iceberg.html" target="_blank">
        <img src="{{site.url}}/assets/images/logos/apache-iceberg.png" title="Iceberg connector">
      </a>
    </div>
  </div>
</div>

Hive is still around, just like HDFS, but we consider them both close to legacy
status. Yet all four connectors could benefit from caching. Good news came at
Trino Summit 2022 when Hope Wang and Beinan Wang from
[Alluxio]({{site.url}}/ecosystem/add-on.html#alluxio) presented about their
integration with Trino and the Hive connector - [Trino optimization with
distributed caching on data lake]({% post_url
2023-07-21-trino-fest-2023-alluxio-recap %}). They mentioned plans to open

> source their implementation and an initial PR was created.

Reviewer: Move `pull request (PR)` up here, as this is the first use of the term.

Author: Done

<div class="container">
  <div class="row">
    <div class="col-sm"></div>
    <div class="col-sm">
      <img src="{{site.url}}/assets/images/logos/alluxio.png" title="Alluxio">
    </div>
    <div class="col-sm"></div>
  </div>
</div>

## Collaboration

> The initial presentation and pull request (PR) planted a seed in the community.

Reviewer: And just go with `PR` down here.

The Trino project had been moving fast in terms of deprecating the old

> dependencies from the Hadoop and Hive ecosystem so the Alluxio PR was not ready

Reviewer: ```suggestion
dependencies from the Hadoop and Hive ecosystem, so the Alluxio PR was not ready
```
Is there a more specific (but still diplomatic) way to say "not ready for a merge?"

Author: Reworded.

for a merge. Discussions with [David Phillips](https://github.com/electrum) laid
out the path to adjust to the new file system support.

In the end it was [Florent Delannoy](https://github.com/pluies) who started
another [PR for file system caching support, specifically for the Delta Lake

> connector](https://github.com/trinodb/trino/pull/18719). His team mate [Jonas

Reviewer: ```suggestion
connector](https://github.com/trinodb/trino/pull/18719). His teammate [Jonas
```

Irgens Kylling](https://github.com/jkylling), also a [presenter from Trino Fest
  2023]({% post_url 2023-07-14-trino-fest-2023-dune %}), took over the work on
the PR. The collaboration on it was an **epic effort**. Over 300
comments directly on GitHub and numerous hours of coding, reviewing, testing,

> and discussion on slack and elsewhere finally ended with a merge after many

Reviewer: ```suggestion
and discussion on Slack and elsewhere finally ended with a merge after many
```
"ended with a merge" feels anticlimactic, maybe specify that this means the launch of the feature and it shipping live with Trino

Author: Reworded

months.

Special props for their help for Florent and Jonas must go out to [David
Phillips](https://github.com/electrum), [Raunaq
Morarka](https://github.com/raunaqmorarka), [Piotr
Findeisen](https://github.com/findepi), [Mateusz
Gajewski](https://github.com/wendigo), [Beinan Wang](https://github.com/beinan),
[Amogh Margoor](https://github.com/amoghmargoor), and [Marton
Bod](https://github.com/marton-bod).

## Finishing

> In parallel to the work on the initial PR for Delta Lake, I ended working on the

Reviewer: ```suggestion
In parallel to the work on the initial PR for Delta Lake, I ended up working on the
```

Reviewer: Maybe say "yours truly" instead of "I," which is a little jarring for a blog that isn't otherwise written in first person.

Author: sounds good

documentation, and pulled together an [issue and conversations to streamline the
roll out](https://github.com/trinodb/trino/issues/20550).

[Mateusz Gajewski](https://github.com/wendigo) had also put together a PR to
remove the old RubiX integration already. With the merge of the initial PR we
were off to the races. We merged the removal of RubiX and the addition of the
docs. Mateusz also added support for OpenTelemetry.

[Manish Malhorta](https://github.com/osscm) and [Amogh
Margoor](https://github.com/amoghmargoor) sent a PR for Iceberg support. They
were also about to add Hive support, when [Raunaq
Morarka](https://github.com/raunaqmorarka) beat them and submitted that PR.

After some final clean up, [Cole Bowden](https://github.com/colebow) and [Martin
Traverso](https://github.com/martint) got the release notes together and shipped
[Trino 439]({{site.url}}/docs/current/release/release-438.html)! Now you can use
it too.

> ## Using file system cache

Reviewer: "the file system cache" or "file system caching"?

There are only a few relatively simple steps to add file system caching to your
catalogs that use Delta Lake, Hive, or Iceberg connectors:

* Provision fast local file system storage on all your Trino cluster nodes. How
  you do that depends on your cluster provisioning.
* Enable file system caching and configure the cache location, for example at
  `/tmp/trino-cache` on the nodes for in your catalog properties files

```
fs.cache.enabled=true
fs.cache.directories=/tmp/trino-cache
```

After a cluster restart, file system caching is active for the configured
catalogs. You can  and you can tweak it with [further, optional configuration
properties]({{site.url}}/docs/current/object-storage/file-system-cache.html).

## What's next

What a success! It took many members from the global Trino village to get this
feature added. Now our users across the globe can enjoy even more benefits of
using Trino, and also participate in our next steps:

* Further improvements to the current implementation, maybe adding worker to

>   worker connections for exchanging cached files.

Reviewer: worker-to-worker
